libwololokingdoms/fixes/queuetechs.cpp

```diff
@@ -0,0 +1,604 @@
+// `effectCommands` does not contain such an effect command, or the final such
+// effect command sets the research time to a nonzero value.
+// @param effectCommands the list of effect commands to check
+// @param tech the technology to check
+bool researchTimeChangesToZero(std::vector<genie::EffectCommand>& effectCommands, int techId) {
+  // Loops backwards, as if multiple Effect Commands set the research time, the
+  // final one in the list is the one that determines the value.
+  for (int i = effectCommands.size() - 1; i != -1; --i) {
+    auto ec = effectCommands[i];
+    if (ec.Type != genie::EffectCommand::EffectType::TechTimeModifier) { continue; }
+    if (ec.TargetUnit != techId) { continue; }
+    return ec.Amount == 0;
+  }
+  return false;
+}
+
+// Sets the discount for the Burmese Monastery technology bonus,
+// the Italian Dock technology bonus, and the Spanish Blacksmith technology
+// bonus.
+// @param df the dat file to modify
+// @param maps a `techId` the the id of the dummy unit that represent it
+void setDiscounts(genie::DatFile * df, unordered_map<int, int>& techIdToUnitId) {
+  for (auto civ : df->Civs) {
+    auto techTreeId = civ.TechTreeID;
+    auto techTreeEffect = &df->Effects[techTreeId];
+    auto numCmdsStart = techTreeEffect->EffectCommands.size();
+    for (int i = 0; i != numCmdsStart; ++i) {
+      auto cmd = &techTreeEffect->EffectCommands[i];
+      // Filters out effect commands that are not tech cost modifiers.
+      if (cmd->Type != genie::EffectCommand::EffectType::TechCostModifier) { continue; }
+
+      auto techId = cmd->TargetUnit;
+      // Filters out the Age technologies for the Italians, since Ages cannot currently be queued
+      if (isAge(techId)) { continue; }
+      // Filters out technologies that are "free" with 0 research time
+      if (&df->Techs[cmd->TargetUnit].ResearchTime == 0
+			|| researchTimeChangesToZero(techTreeEffect->EffectCommands, cmd->TargetUnit)) {
+        continue;
+      }
+
+      auto cmdNew = genie::EffectCommand();
+
+      cmdNew.Type = cmd->AttributeID == MODE_SET_COST
+              ? (int) genie::EffectCommand::EffectType::AbsoluteAttributeModifier
+              : (int) genie::EffectCommand::EffectType::RelativeAttributeModifier;
+      cmdNew.TargetUnit = techIdToUnitId[techId];
+      switch (cmd->UnitClassID) {
+      case TECH_FOOD_COST:
+        cmdNew.AttributeID = genie::EffectCommand::Attributes::FoodCosts;
+        break;
+      case TECH_WOOD_COST:
+        cmdNew.AttributeID = genie::EffectCommand::Attributes::WoodCosts;
+        break;
+      case TECH_STONE_COST:
+        cmdNew.AttributeID = genie::EffectCommand::Attributes::StoneCosts;
+        break;
+      case TECH_GOLD_COST:
+        cmdNew.AttributeID = genie::EffectCommand::Attributes::GoldCosts;
+        break;
+      default:
+        assert(false, "Unit Class ID must be between 0 and 3, inclusive, to "
+                      "represent a resource cost.");
```

Comment:
C-style `assert()` only takes a single argument, sometimes people do:
```c
    assert(!"Unit Class ID must be between 0 and 3, inclusive, to represent a resource cost.");
```
which is horrible but works.

This file also needs an `#include <cassert>` at the top to make the `assert()` macro available (it might be pulled in by something else in some standard library implementations, but it's not in my Clang version)

Reply:
Added the `include` statement.
I'll just leave that line as `assert(false);` and refactor the string to a comment.